tenant/error.go

```
@@ -0,0 +1,56 @@
+package tenant
+
+import (
+	"regexp"
+
+	"github.com/giantswarm/microerror"
+)
+
+var (
+	APINotAvailablePatterns = []*regexp.Regexp{
+		// A regular expression representing DNS errors for the tenant API domain.
+		regexp.MustCompile(`dial tcp: lookup .* on .*:53: (no such host|server misbehaving)`),
+		// A regular expression representing EOF errors for the tenant API domain.
+		regexp.MustCompile(`[Get|Patch|Post] https://api\..*/api/v1/nodes.* (unexpected )?EOF`),
+		// A regular expression representing EOF errors for the tenant API domain.
+		regexp.MustCompile(`[Get|Patch|Post] https://api\..*/api/v1/namespaces/*/.* (unexpected )?EOF`),
+		// A regular expression representing TLS errors related to establishing
+		// connections to tenant clusters while the tenant API is not fully up.
+		regexp.MustCompile(`[Get|Patch|Post] https://api\..*/api/v1/nodes.* net/http: (TLS handshake timeout|request canceled).*?`),
+		// A regular expression representing timeout errors related to establishing
+		// TCP connections to tenant clusters while the tenant API is not fully up.
+		regexp.MustCompile(`[Get|Patch|Post] https://api\..* dial tcp .* i/o timeout`),
+		// A regular expression representing the kind of transient errors related to
+		// certificates returned while the tenant API is not fully up.
+		regexp.MustCompile(`[Get|Patch|Post] https://api\..*: x509: (certificate is valid for ingress.local, not api\..*|certificate signed by unknown authority \(possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate.*?\))`),
+	}
```

When we group different kind of networking errors here, they get effectively swallowed and hidden by calling component. There can be many different reasons why these errors occur, but I'm wondering if we loose observability with this kind of functionality?

My understanding on usage of this is something like following:

```go
var err error
var tc *tenantcluster.TenantCluster
{
    c := tenantcluster.Config{}
    tc, err = tenantcluster.New(c)
    if err != nil {
        return microerror.Mask(err)
    }
}

//  ....

var tenantK8sClient kubernetes.Interface
{
    tenantK8sClient, err = tc.NewK8sClient(ctx, "abcde", "foobar.com")
    if tenant.IsAPINotAvailable(err) {
        // Fine, wait a while and retry.
        reconciliationcanceledcontext.SetCanceled(ctx)
        return nil
    } else if err != nil {
        // escalate error
        return microerror.Mask(err)
    }
}

// continue...
```

How can we now notice problems between control plane and tenant cluster? Would it make sense to think about possibility of exposing metrics for specific errors so that our monitoring could catch these while we keep the application code simple and clean?

---

That sounds like a good topic for SIG Monitoring. Let's discuss this and see what we can do. Here I only copied what we have. There is nothing new. I agree with you and think having more metrics or logging makes sense.

---

I created a ticket there: https://github.com/giantswarm/giantswarm/issues/5494

---

Thank you.